// @tamagui/label/dist/jsx/Label.js
import { useComposedRefs } from "@tamagui/compose-refs";
import { createContext } from "@tamagui/create-context";
import { focusFocusable } from "@tamagui/focusable";
import { getButtonSized } from "@tamagui/get-button-sized";
import { getFontSized } from "@tamagui/get-font-sized";
import { SizableText } from "@tamagui/text";
import {
  isWeb,
  styled,
  themeable
} from "@tamagui/web";
import * as React from "react";
const NAME = "Label";
const [LabelProvider, useLabelContextImpl] = createContext(NAME, {
  id: void 0,
  controlRef: { current: null }
});
const LabelFrame = styled(SizableText, {
  name: "Label",
  tag: "label",
  variants: {
    unstyled: {
      false: {
        size: "$true",
        color: "$color",
        backgroundColor: "transparent",
        display: "flex",
        alignItems: "center",
        userSelect: "none",
        cursor: "default",
        pressStyle: {
          color: "$colorPress"
        }
      }
    },
    size: {
      "...size": (val, extras) => {
        const buttonStyle = getButtonSized(val, extras);
        const buttonHeight = buttonStyle?.height;
        const fontStyle = getFontSized(val, extras);
        return {
          ...fontStyle,
          lineHeight: buttonHeight ? extras.tokens.size[buttonHeight] : void 0
        };
      }
    }
  },
  defaultVariants: {
    unstyled: false
  }
});
const LabelComponent = React.forwardRef(
  (props, forwardedRef) => {
    const { htmlFor, id: idProp, ...labelProps } = props;
    const controlRef = React.useRef(null);
    const ref = React.useRef(null);
    const composedRefs = useComposedRefs(forwardedRef, ref);
    const backupId = React.useId();
    const id = idProp ?? backupId;
    if (isWeb) {
      React.useEffect(() => {
        if (htmlFor) {
          const element = document.getElementById(htmlFor);
          const label = ref.current;
          if (label && element) {
            const getAriaLabel = () => element.getAttribute("aria-labelledby");
            const ariaLabelledBy = [id, getAriaLabel()].filter(Boolean).join(" ");
            element.setAttribute("aria-labelledby", ariaLabelledBy);
            controlRef.current = element;
            return () => {
              if (!id)
                return;
              const ariaLabelledBy2 = getAriaLabel()?.replace(id, "");
              if (ariaLabelledBy2 === "") {
                element.removeAttribute("aria-labelledby");
              } else if (ariaLabelledBy2) {
                element.setAttribute("aria-labelledby", ariaLabelledBy2);
              }
            };
          }
        }
      }, [id, htmlFor]);
    }
    return <LabelProvider id={id} controlRef={controlRef}><LabelFrame
      id={id}
      htmlFor={htmlFor}
      {...labelProps}
      ref={composedRefs}
      onMouseDown={(event) => {
        props.onMouseDown?.(event);
        if (!event.defaultPrevented && event.detail > 1) {
          event.preventDefault();
        }
      }}
      onPress={(event) => {
        props.onPress?.(event);
        if (isWeb) {
          if (htmlFor || !controlRef.current || event.defaultPrevented)
            return;
          const isClickingControl = controlRef.current.contains(
            event.target
          );
          const isUserClick = event.isTrusted === true;
          if (!isClickingControl && isUserClick) {
            controlRef.current.click();
            controlRef.current.focus();
          }
        } else {
          if (props.htmlFor) {
            focusFocusable(props.htmlFor);
          }
        }
      }}
    /></LabelProvider>;
  }
);
LabelComponent.displayName = NAME;
const Label = LabelFrame.extractable(themeable(LabelComponent), {
  neverFlatten: true
});
const useLabelContext = (element) => {
  const context = useLabelContextImpl("LabelConsumer");
  const { controlRef } = context;
  React.useEffect(() => {
    if (element)
      controlRef.current = element;
  }, [element, controlRef]);
  return context.id;
};
const Root = Label;
export {
  Label,
  LabelFrame,
  useLabelContext
};
