// @tamagui/next-theme/dist/cjs/NextThemeProvider.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NextThemeProvider_exports = {};
__export(NextThemeProvider_exports, {
  NextThemeProvider: () => NextThemeProvider
});
module.exports = __toCommonJS(NextThemeProvider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_use_event = require("@tamagui/use-event");
var import_head = __toESM(require("next/head"));
var React = __toESM(require("react"));
var import_react = require("react");
var import_constants = require("./constants");
var import_helpers = require("./helpers");
var import_ThemeSettingContext = require("./ThemeSettingContext");
var import_useIsomorphicLayoutEffect = require("./useIsomorphicLayoutEffect");
const NextThemeProvider = (0, import_react.memo)(
  ({
    forcedTheme,
    disableTransitionOnChange = true,
    enableSystem = true,
    enableColorScheme = true,
    storageKey = "theme",
    themes = import_constants.colorSchemes,
    defaultTheme = enableSystem ? "system" : "light",
    attribute = "class",
    skipNextHead,
    onChangeTheme,
    value = {
      dark: "t_dark",
      light: "t_light"
    },
    children
  }) => {
    const [theme, setThemeState] = (0, import_react.useState)(() => (0, import_helpers.getTheme)(storageKey, defaultTheme));
    const [resolvedTheme, setResolvedTheme] = (0, import_react.useState)(() => (0, import_helpers.getTheme)(storageKey));
    const attrs = !value ? themes : Object.values(value);
    const handleMediaQuery = (0, import_use_event.useEvent)((e) => {
      const systemTheme2 = (0, import_helpers.getSystemTheme)(e);
      React.startTransition(() => {
        setResolvedTheme(systemTheme2);
      });
      if (theme === "system" && !forcedTheme) {
        handleChangeTheme(systemTheme2, false);
      }
    });
    const mediaListener = (0, import_react.useRef)(handleMediaQuery);
    mediaListener.current = handleMediaQuery;
    const handleChangeTheme = (0, import_use_event.useEvent)(
      (theme2, updateStorage = true, updateDOM = true) => {
        let name = (value == null ? void 0 : value[theme2]) || theme2;
        if (updateStorage) {
          try {
            localStorage.setItem(storageKey, theme2);
          } catch (e) {
          }
        }
        if (theme2 === "system" && enableSystem) {
          const resolved = (0, import_helpers.getSystemTheme)();
          name = (value == null ? void 0 : value[resolved]) || resolved;
        }
        onChangeTheme == null ? void 0 : onChangeTheme(name.replace("t_", ""));
        if (updateDOM) {
          const d = document.documentElement;
          if (attribute === "class") {
            d.classList.remove(...attrs);
            d.classList.add(name);
          } else {
            d.setAttribute(attribute, name);
          }
        }
      }
    );
    (0, import_useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(() => {
      const handler = (...args) => mediaListener.current(...args);
      const media = window.matchMedia(import_constants.MEDIA);
      media.addListener(handler);
      handler(media);
      return () => {
        media.removeListener(handler);
      };
    }, []);
    const set = (0, import_use_event.useEvent)((newTheme) => {
      if (forcedTheme) {
        handleChangeTheme(newTheme, true, false);
      } else {
        handleChangeTheme(newTheme);
      }
      setThemeState(newTheme);
    });
    (0, import_react.useEffect)(() => {
      const handleStorage = (e) => {
        if (e.key !== storageKey) {
          return;
        }
        const theme2 = e.newValue || defaultTheme;
        set(theme2);
      };
      window.addEventListener("storage", handleStorage);
      return () => {
        window.removeEventListener("storage", handleStorage);
      };
    }, [defaultTheme, set, storageKey]);
    (0, import_useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(() => {
      if (!enableColorScheme)
        return;
      const colorScheme = (
        // If theme is forced to light or dark, use that
        forcedTheme && import_constants.colorSchemes.includes(forcedTheme) ? forcedTheme : (
          // If regular theme is light or dark
          theme && import_constants.colorSchemes.includes(theme) ? theme : (
            // If theme is system, use the resolved version
            theme === "system" ? resolvedTheme || null : null
          )
        )
      );
      const userPrefers = typeof window !== "undefined" && window.matchMedia && window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light";
      const wePrefer = colorScheme || "light";
      if (userPrefers !== wePrefer) {
        document.documentElement.style.setProperty("color-scheme", colorScheme);
      }
    }, [enableColorScheme, theme, resolvedTheme, forcedTheme]);
    const toggle = (0, import_use_event.useEvent)(() => {
      const order = resolvedTheme === "dark" ? ["system", "light", "dark"] : ["system", "dark", "light"];
      const next = order[(order.indexOf(theme) + 1) % order.length];
      set(next);
    });
    const contextResolvedTheme = theme === "system" ? resolvedTheme : theme;
    const systemTheme = enableSystem ? resolvedTheme : void 0;
    const contextValue = (0, import_react.useMemo)(() => {
      const value2 = {
        theme,
        current: theme,
        set,
        toggle,
        forcedTheme,
        resolvedTheme: contextResolvedTheme,
        themes: enableSystem ? [...themes, "system"] : themes,
        systemTheme
      };
      return value2;
    }, [
      theme,
      set,
      toggle,
      forcedTheme,
      contextResolvedTheme,
      enableSystem,
      themes,
      systemTheme
    ]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ThemeSettingContext.ThemeSettingContext.Provider, { value: contextValue, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        ThemeScript,
        {
          ...{
            forcedTheme,
            storageKey,
            systemTheme: resolvedTheme,
            attribute,
            value,
            enableSystem,
            defaultTheme,
            attrs,
            skipNextHead
          }
        }
      ),
      (0, import_react.useMemo)(() => children, [children])
    ] });
  }
);
const ThemeScript = (0, import_react.memo)(
  ({
    forcedTheme,
    storageKey,
    attribute,
    enableSystem,
    defaultTheme,
    value,
    attrs,
    skipNextHead
  }) => {
    const optimization = (() => {
      if (attribute === "class") {
        const removeClasses = attrs.map((t) => `d.remove('${t}')`).join(";");
        return `var d=document.documentElement.classList;${removeClasses};`;
      } else {
        return `var d=document.documentElement;`;
      }
    })();
    const updateDOM = (name, literal) => {
      name = (value == null ? void 0 : value[name]) || name;
      const val = literal ? name : `'${name}'`;
      if (attribute === "class") {
        return `d.add(${val})`;
      }
      return `d.setAttribute('${attribute}', ${val})`;
    };
    const defaultSystem = defaultTheme === "system";
    const contents = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: forcedTheme ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "script",
      {
        dangerouslySetInnerHTML: {
          // These are minified via Terser and then updated by hand, don't recommend
          __html: `!function(){${optimization}${updateDOM(forcedTheme)}}()`
        }
      },
      "next-themes-script"
    ) : enableSystem ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "script",
      {
        dangerouslySetInnerHTML: {
          __html: `!function(){try {${optimization}var e=localStorage.getItem('${storageKey}');${!defaultSystem ? updateDOM(defaultTheme) + ";" : ""}if("system"===e||(!e&&${defaultSystem})){var t="${import_constants.MEDIA}",m=window.matchMedia(t);m.media!==t||m.matches?${updateDOM(
            "dark"
          )}:${updateDOM("light")}}else if(e) ${value ? `var x=${JSON.stringify(value)};` : ""}${updateDOM(value ? "x[e]" : "e", true)}}catch(e){}}()`
        }
      },
      "next-themes-script"
    ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "script",
      {
        dangerouslySetInnerHTML: {
          __html: `!function(){try{${optimization}var e=localStorage.getItem("${storageKey}");if(e){${value ? `var x=${JSON.stringify(value)};` : ""}${updateDOM(value ? "x[e]" : "e", true)}}else{${updateDOM(
            defaultTheme
          )};}}catch(t){}}();`
        }
      },
      "next-themes-script"
    ) });
    if (skipNextHead)
      return contents;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_head.default, { children: contents });
  },
  (prevProps, nextProps) => {
    if (prevProps.forcedTheme !== nextProps.forcedTheme)
      return false;
    return true;
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  NextThemeProvider
});
