// @tamagui/web/dist/cjs/helpers/expandStyles.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var expandStyles_exports = {};
__export(expandStyles_exports, {
  expandStylesAndRemoveNullishValues: () => expandStylesAndRemoveNullishValues,
  fixStyles: () => fixStyles
});
module.exports = __toCommonJS(expandStyles_exports);
var import_constants = require("@tamagui/constants");
var import_expandStyle = require("./expandStyle");
var import_normalizeShadow = require("./normalizeShadow");
var import_normalizeValueWithProperty = require("./normalizeValueWithProperty");
var import_pseudoDescriptors = require("./pseudoDescriptors");
function expandStylesAndRemoveNullishValues(style) {
  const res = {};
  for (let key in style) {
    const valIn = style[key];
    if (valIn == null)
      continue;
    if (key in import_pseudoDescriptors.pseudoDescriptors) {
      res[key] = expandStylesAndRemoveNullishValues(valIn);
      continue;
    }
    const val = (0, import_normalizeValueWithProperty.normalizeValueWithProperty)(valIn, key);
    const out = (0, import_expandStyle.expandStyle)(key, val);
    if (out) {
      Object.assign(res, Object.fromEntries(out));
    } else {
      res[key] = val;
    }
  }
  fixStyles(res);
  return res;
}
function fixStyles(style) {
  if (process.env.TAMAGUI_TARGET === "native") {
    if ("elevationAndroid" in style) {
      style["elevation"] = style.elevationAndroid;
      delete style.elevationAndroid;
    }
  }
  if (style.shadowRadius || style.shadowColor || style.shadowOpacity || style.shadowOffset) {
    Object.assign(style, (0, import_normalizeShadow.normalizeShadow)(style));
  }
  for (const key in borderDefaults) {
    if (key in style && !style[borderDefaults[key]]) {
      style[borderDefaults[key]] = "solid";
    }
  }
}
const nativeStyle = import_constants.isWeb ? null : "borderStyle";
const borderDefaults = {
  borderWidth: "borderStyle",
  borderBottomWidth: nativeStyle || "borderBottomStyle",
  borderTopWidth: nativeStyle || "borderTopStyle",
  borderLeftWidth: nativeStyle || "borderLeftStyle",
  borderRightWidth: nativeStyle || "borderRightStyle"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  expandStylesAndRemoveNullishValues,
  fixStyles
});
