// @tamagui/static/dist/esm/extractor/loadTamagui.js
import { basename, dirname, extname, join, relative, resolve } from "path";
import { Color, colorLog } from "@tamagui/cli-color";
import { getDefaultTamaguiConfig } from "@tamagui/config-default-node";
import { createTamagui } from "@tamagui/core-node";
import esbuild from "esbuild";
import { existsSync, pathExists, readJSON, writeFile } from "fs-extra";
import { SHOULD_DEBUG } from "../constants";
import { getNameToPaths, registerRequire } from "../require";
import {
  TamaguiProjectInfo,
  getBundledConfig,
  hasBundledConfigChanged,
  loadComponents
} from "./bundleConfig";
import {
  generateTamaguiStudioConfig,
  generateTamaguiStudioConfigSync,
  generateTamaguiThemes
} from "./generateTamaguiStudioConfig";
import { getTamaguiConfigPathFromOptionsConfig } from "./getTamaguiConfigPathFromOptionsConfig";
const getFilledOptions = (propsIn) => ({
  // defaults
  config: "tamagui.config.ts",
  components: ["tamagui"],
  ...propsIn
});
async function loadTamagui(propsIn) {
  const options = getFilledOptions(propsIn);
  await generateThemesAndLog(options);
  const bundleInfo = await getBundledConfig(options);
  if (!bundleInfo) {
    console.warn(
      `No bundled config generated, maybe an error in bundling. Set DEBUG=tamagui and re-run to get logs.`
    );
    return null;
  }
  if (!hasBundledConfigChanged()) {
    return bundleInfo;
  }
  await generateTamaguiStudioConfig(options, bundleInfo);
  if (bundleInfo) {
    const config = createTamagui(bundleInfo.tamaguiConfig);
    if (options.outputCSS) {
      colorLog(Color.FgYellow, `    \u27A1 [tamagui] outputCSS: ${options.outputCSS}
`);
      await writeFile(options.outputCSS, config.getCSS());
    }
  }
  return bundleInfo;
}
let waiting = false;
let hasLoggedOnce = false;
const generateThemesAndLog = async (options) => {
  if (waiting)
    return;
  if (!options.themeBuilder)
    return;
  try {
    waiting = true;
    await new Promise((res) => setTimeout(res, 30));
    const didGenerate = await generateTamaguiThemes(options);
    if (!hasLoggedOnce || didGenerate) {
      hasLoggedOnce = true;
      const whitespaceBefore = `    `;
      colorLog(Color.FgYellow, `${whitespaceBefore}\u27A1 [tamagui] Generated themes:`);
      colorLog(
        Color.Dim,
        `
${whitespaceBefore}${relative(process.cwd(), options.themeBuilder.output)}`
      );
    }
  } finally {
    waiting = false;
  }
};
function loadTamaguiSync(propsIn) {
  const props = getFilledOptions(propsIn);
  const unregisterRequire = registerRequire();
  try {
    process.env.IS_STATIC = "is_static";
    process.env.TAMAGUI_IS_SERVER = "true";
    const devValueOG = globalThis["__DEV__"];
    globalThis["__DEV__"] = process.env.NODE_ENV === "development";
    try {
      let tamaguiConfig = null;
      if (props.config) {
        const configPath = getTamaguiConfigPathFromOptionsConfig(props.config);
        const exp = require(configPath);
        tamaguiConfig = exp["default"] || exp;
        if (!tamaguiConfig || !tamaguiConfig.parsed) {
          const confPath = require.resolve(configPath);
          throw new Error(`Can't find valid config in ${confPath}:
          
  Be sure you "export default" the config.`);
        }
      }
      const components = loadComponents(props);
      if (!components) {
        throw new Error(`No components loaded`);
      }
      if (process.env.DEBUG === "tamagui") {
        console.log(`components`, components);
      }
      process.env.IS_STATIC = void 0;
      globalThis["__DEV__"] = devValueOG;
      if (props.config && tamaguiConfig) {
        createTamagui(tamaguiConfig);
      }
      const info = {
        components,
        tamaguiConfig,
        nameToPaths: getNameToPaths()
      };
      generateTamaguiStudioConfigSync(props, info);
      return info;
    } catch (err) {
      if (err instanceof Error) {
        console.warn(
          `Error loading tamagui.config.ts (set DEBUG=tamagui to see full stack), running tamagui without custom config`
        );
        console.log(`

    ${err.message}

`);
        if (SHOULD_DEBUG) {
          console.error(err.stack);
        }
      } else {
        console.error(`Error loading tamagui.config.ts`, err);
      }
      return {
        components: [],
        tamaguiConfig: createTamagui(getDefaultTamaguiConfig()),
        nameToPaths: {}
      };
    }
  } finally {
    unregisterRequire();
  }
}
async function getOptions({
  root = process.cwd(),
  tsconfigPath = "tsconfig.json",
  tamaguiOptions,
  host,
  debug
} = {}) {
  const dotDir = join(root, ".tamagui");
  let pkgJson = {};
  try {
    pkgJson = await readJSON(join(root, "package.json"));
  } catch (err) {
  }
  return {
    mode: process.env.NODE_ENV === "production" ? "production" : "development",
    root,
    host: host || "127.0.0.1",
    pkgJson,
    debug,
    tsconfigPath,
    tamaguiOptions: {
      components: ["tamagui"],
      ...tamaguiOptions,
      config: await getDefaultTamaguiConfigPath(root, tamaguiOptions == null ? void 0 : tamaguiOptions.config)
    },
    paths: {
      dotDir,
      conf: join(dotDir, "tamagui.config.json"),
      types: join(dotDir, "types.json")
    }
  };
}
function resolveWebOrNativeSpecificEntry(entry) {
  const workspaceRoot = resolve();
  const resolved = require.resolve(entry, { paths: [workspaceRoot] });
  const ext = extname(resolved);
  const fileName = basename(resolved).replace(ext, "");
  const specificExt = process.env.TAMAGUI_TARGET === "web" ? "web" : "native";
  const specificFile = join(dirname(resolved), fileName + "." + specificExt + ext);
  if (existsSync(specificFile)) {
    return specificFile;
  }
  return entry;
}
const defaultPaths = ["tamagui.config.ts", join("src", "tamagui.config.ts")];
let hasWarnedOnce = false;
async function getDefaultTamaguiConfigPath(root, configPath) {
  const searchPaths = [
    ...new Set(
      [configPath, ...defaultPaths].filter(Boolean).map((p) => join(root, p))
    )
  ];
  for (const path of searchPaths) {
    if (await pathExists(path)) {
      return path;
    }
  }
  if (!hasWarnedOnce) {
    hasWarnedOnce = true;
    console.warn(`Warning: couldn't find tamagui.config.ts in the following paths given configuration "${configPath}":
    ${searchPaths.join(`
  `)}
  `);
  }
}
async function watchTamaguiConfig(tamaguiOptions) {
  var _a;
  const options = await getOptions({ tamaguiOptions });
  if (!options.tamaguiOptions.config) {
    throw new Error(`No config`);
  }
  if (process.env.NODE_ENV === "production") {
    return {
      dispose() {
      }
    };
  }
  const disposeConfigWatcher = await esbuildWatchFiles(
    options.tamaguiOptions.config,
    () => {
      void generateTamaguiStudioConfig(options.tamaguiOptions, null, true);
    }
  );
  const themeBuilderInput = (_a = options.tamaguiOptions.themeBuilder) == null ? void 0 : _a.input;
  let disposeThemesWatcher;
  if (themeBuilderInput) {
    let inputPath = themeBuilderInput;
    try {
      inputPath = require.resolve(themeBuilderInput);
    } catch {
    }
    disposeThemesWatcher = await esbuildWatchFiles(inputPath, () => {
      void generateThemesAndLog(options.tamaguiOptions);
    });
  }
  return {
    dispose() {
      disposeConfigWatcher();
      disposeThemesWatcher == null ? void 0 : disposeThemesWatcher();
    }
  };
}
async function esbuildWatchFiles(entry, onChanged) {
  let hasRunOnce = false;
  const context = await esbuild.context({
    bundle: true,
    entryPoints: [entry],
    resolveExtensions: [".ts", ".tsx", ".js", ".jsx", ".mjs"],
    logLevel: "silent",
    write: false,
    plugins: [
      {
        name: `on-rebuild`,
        setup({ onEnd }) {
          onEnd(() => {
            if (!hasRunOnce) {
              hasRunOnce = true;
            } else {
              onChanged();
            }
          });
        }
      }
    ]
  });
  void context.watch();
  return () => {
    context.dispose();
  };
}
export {
  TamaguiProjectInfo,
  getOptions,
  loadTamagui,
  loadTamaguiSync,
  resolveWebOrNativeSpecificEntry,
  watchTamaguiConfig
};
