// @tamagui/web/dist/cjs/views/ThemeDebug.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ThemeDebug_exports = {};
__export(ThemeDebug_exports, {
  ThemeDebug: () => ThemeDebug
});
module.exports = __toCommonJS(ThemeDebug_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_use_force_update = require("@tamagui/use-force-update");
var import_react = require("react");
var import_react_dom = require("react-dom");
let node;
function ThemeDebug({
  themeState,
  themeProps,
  children
}) {
  var _a, _b, _c, _d, _e, _f;
  if (themeProps["disable-child-theme"]) {
    return children;
  }
  if (process.env.NODE_ENV === "development") {
    const [onChangeCount, setOnChangeCount] = (0, import_react.useState)(0);
    const rerender = (0, import_use_force_update.useForceUpdate)();
    const id = (0, import_react.useId)();
    if (process.env.NODE_ENV === "development" && typeof document !== "undefined") {
      if (!node) {
        node = document.createElement("div");
        node.style.height = "200px";
        node.style.overflowY = "scroll";
        node.style.position = "fixed";
        node.style.zIndex = 1e7;
        node.style.bottom = "30px";
        node.style.left = "30px";
        node.style.right = "30px";
        node.style.display = "flex";
        node.style.border = "1px solid #888";
        node.style.flexDirection = "row";
        node.style.background = "var(--background)";
        document.body.appendChild(node);
      }
    }
    (0, import_react.useEffect)(() => {
      var _a2, _b2;
      (_b2 = (_a2 = themeState.themeManager) == null ? void 0 : _a2.parentManager) == null ? void 0 : _b2.onChangeTheme((name, manager) => {
        var _a3, _b3, _c2;
        setOnChangeCount((p) => ++p);
        console.warn(
          `theme changed for ${(_a3 = themeState.themeManager) == null ? void 0 : _a3.id} from parent ${(_c2 = (_b3 = themeState.themeManager) == null ? void 0 : _b3.parentManager) == null ? void 0 : _c2.id} to new name`,
          name
        );
      });
    }, [themeState.themeManager]);
    (0, import_react.useEffect)(() => {
      const tm = setInterval(rerender, 1e3);
      return () => clearTimeout(tm);
    }, []);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      (0, import_react_dom.createPortal)(
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          "code",
          {
            style: {
              whiteSpace: "pre"
            },
            children: [
              "<Theme ",
              id,
              " />\xA0",
              JSON.stringify(
                {
                  name: themeState.state.name,
                  className: themeState.state.className,
                  inverse: themeProps.inverse,
                  forceClassName: themeProps.forceClassName,
                  parent: (_a = themeState.themeManager) == null ? void 0 : _a.state.parentName,
                  id: (_b = themeState.themeManager) == null ? void 0 : _b.id,
                  parentId: (_d = (_c = themeState.themeManager) == null ? void 0 : _c.parentManager) == null ? void 0 : _d.id,
                  isNew: themeState.isNewTheme,
                  onChangeCount,
                  listening: [...((_e = themeState.themeManager) == null ? void 0 : _e["_listeningIds"]) || []].join(
                    ","
                  ),
                  _numChangeEventsSent: (_f = themeState.themeManager) == null ? void 0 : _f["_numChangeEventsSent"]
                },
                null,
                2
              )
            ]
          }
        ),
        node
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { style: { color: "red" }, children: id }),
      children
    ] });
  }
  return children;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ThemeDebug
});
