// @tamagui/select/dist/esm/context.js
import { jsx } from "react/jsx-runtime";
import { createContextScope } from "@tamagui/create-context";
import { SELECT_NAME } from "./constants";
const [createSelectContext, createSelectScope] = createContextScope(SELECT_NAME);
const [SelectProvider, useSelectContext] = createSelectContext(SELECT_NAME);
const [createSelectItemParentContext, createSelectItemParentScope] = createContextScope(SELECT_NAME);
const [SelectItemParentProvider, useSelectItemParentContext] = createSelectContext(SELECT_NAME);
const ForwardSelectContext = ({
  __scopeSelect,
  context,
  itemContext,
  children
}) => {
  return /* @__PURE__ */ jsx(SelectProvider, { isInSheet: true, scope: __scopeSelect, ...context, children: /* @__PURE__ */ jsx(SelectItemParentProvider, { scope: __scopeSelect, ...itemContext, children }) });
};
export {
  ForwardSelectContext,
  SelectItemParentProvider,
  SelectProvider,
  createSelectContext,
  createSelectItemParentContext,
  createSelectItemParentScope,
  createSelectScope,
  useSelectContext,
  useSelectItemParentContext
};
