// @tamagui/sheet/dist/jsx/useSheetChildren.mjs
import React, { isValidElement } from "react";
function useSheetChildren(childrenProp) {
  let handleComponent = null;
  let overlayComponent = null;
  let frameComponent = null;
  React.Children.forEach(childrenProp, (child) => {
    if (isValidElement(child)) {
      const name = child.type?.["staticConfig"]?.componentName;
      switch (name) {
        case "SheetHandle":
          handleComponent = child;
          break;
        case "Sheet":
          frameComponent = child;
          break;
        case "SheetOverlay":
          overlayComponent = child;
          break;
        default:
          console.warn("Warning: passed invalid child to Sheet", child);
      }
    }
  });
  let bottomCoverComponent = null;
  if (frameComponent) {
    bottomCoverComponent = React.cloneElement(frameComponent, {
      children: null,
      position: "absolute",
      bottom: -100,
      height: 110,
      left: 0,
      right: 0
    });
  }
  return {
    handleComponent,
    overlayComponent,
    frameComponent,
    bottomCoverComponent
  };
}
export {
  useSheetChildren
};
