// @tamagui/select/dist/cjs/Select.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Select_exports = {};
__export(Select_exports, {
  Select: () => Select,
  SelectGroupFrame: () => SelectGroupFrame,
  SelectIcon: () => SelectIcon,
  SelectItemTextFrame: () => SelectItemTextFrame,
  SelectSeparator: () => SelectSeparator
});
module.exports = __toCommonJS(Select_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_adapt = require("@tamagui/adapt");
var import_compose_refs = require("@tamagui/compose-refs");
var import_core = require("@tamagui/core");
var import_get_token = require("@tamagui/get-token");
var import_list_item = require("@tamagui/list-item");
var import_portal = require("@tamagui/portal");
var import_separator = require("@tamagui/separator");
var import_sheet = require("@tamagui/sheet");
var import_stacks = require("@tamagui/stacks");
var import_text = require("@tamagui/text");
var import_use_controllable_state = require("@tamagui/use-controllable-state");
var React = __toESM(require("react"));
var import_constants = require("./constants");
var import_context = require("./context");
var import_SelectContent = require("./SelectContent");
var import_SelectImpl = require("./SelectImpl");
var import_SelectItem = require("./SelectItem");
var import_SelectScrollButton = require("./SelectScrollButton");
var import_SelectTrigger = require("./SelectTrigger");
var import_SelectViewport = require("./SelectViewport");
var import_useSelectBreakpointActive = require("./useSelectBreakpointActive");
const VALUE_NAME = "SelectValue";
const SelectValueFrame = (0, import_core.styled)(import_text.SizableText, {
  name: VALUE_NAME,
  userSelect: "none"
});
const SelectValue = SelectValueFrame.styleable(function SelectValue2({
  __scopeSelect,
  children: childrenProp,
  placeholder,
  ...props
}, forwardedRef) {
  const context = (0, import_context.useSelectContext)(VALUE_NAME, __scopeSelect);
  const itemParentContext = (0, import_context.useSelectItemParentContext)(VALUE_NAME, __scopeSelect);
  const composedRefs = (0, import_compose_refs.useComposedRefs)(forwardedRef, context.onValueNodeChange);
  const children = childrenProp ?? context.selectedItem;
  const isEmptyValue = context.value == null || context.value === "";
  const selectValueChildren = isEmptyValue ? placeholder ?? children : children;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    SelectValueFrame,
    {
      ...!props.unstyled && {
        size: itemParentContext.size,
        ellipse: true
      },
      ref: composedRefs,
      pointerEvents: "none",
      ...props,
      children: unwrapSelectItem(selectValueChildren)
    }
  );
});
function unwrapSelectItem(selectValueChildren) {
  return React.Children.map(selectValueChildren, (child) => {
    var _a, _b, _c;
    if (child) {
      if (((_b = (_a = child.type) == null ? void 0 : _a.staticConfig) == null ? void 0 : _b.componentName) === ITEM_TEXT_NAME) {
        return child.props.children;
      }
      if ((_c = child.props) == null ? void 0 : _c.children) {
        return unwrapSelectItem(child.props.children);
      }
    }
    return child;
  });
}
const SelectIcon = (0, import_core.styled)(import_stacks.XStack, {
  name: "SelectIcon",
  // @ts-ignore
  "aria-hidden": true,
  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_text.Paragraph, { children: "\u25BC" })
});
const ITEM_TEXT_NAME = "SelectItemText";
const SelectItemTextFrame = (0, import_core.styled)(import_text.SizableText, {
  name: ITEM_TEXT_NAME,
  variants: {
    unstyled: {
      false: {
        userSelect: "none",
        color: "$color",
        ellipse: true
      }
    }
  },
  defaultVariants: {
    unstyled: false
  }
});
const SelectItemText = React.forwardRef(
  (props, forwardedRef) => {
    const { __scopeSelect, className, ...itemTextProps } = props;
    const context = (0, import_context.useSelectContext)(ITEM_TEXT_NAME, __scopeSelect);
    const itemParentContext = (0, import_context.useSelectItemParentContext)(ITEM_TEXT_NAME, __scopeSelect);
    const ref = React.useRef(null);
    const composedRefs = (0, import_compose_refs.useComposedRefs)(forwardedRef, ref);
    const itemContext = (0, import_SelectItem.useSelectItemContext)(ITEM_TEXT_NAME, __scopeSelect);
    const contents = React.useMemo(
      () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        SelectItemTextFrame,
        {
          className,
          size: itemParentContext.size,
          id: itemContext.textId,
          ...itemTextProps,
          ref: composedRefs
        }
      ),
      // eslint-disable-next-line react-hooks/exhaustive-deps
      [props, itemParentContext.size, className, itemContext.textId]
    );
    React.useEffect(() => {
      if (itemParentContext.initialValue === itemContext.value && !context.selectedIndex) {
        context.setSelectedItem(contents);
      }
    }, []);
    React.useEffect(() => {
      return itemParentContext.valueSubscribe((val) => {
        if (val === itemContext.value) {
          context.setSelectedItem(contents);
        }
      });
    }, [itemContext.value]);
    if (itemParentContext.shouldRenderWebNative) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: props.children });
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: contents });
  }
);
SelectItemText.displayName = ITEM_TEXT_NAME;
const ITEM_INDICATOR_NAME = "SelectItemIndicator";
const SelectItemIndicatorFrame = (0, import_core.styled)(import_stacks.XStack, {
  name: ITEM_TEXT_NAME
});
const SelectItemIndicator = React.forwardRef(
  (props, forwardedRef) => {
    const { __scopeSelect, ...itemIndicatorProps } = props;
    const context = (0, import_context.useSelectItemParentContext)(ITEM_INDICATOR_NAME, __scopeSelect);
    const itemContext = (0, import_SelectItem.useSelectItemContext)(ITEM_INDICATOR_NAME, __scopeSelect);
    if (context.shouldRenderWebNative) {
      return null;
    }
    return itemContext.isSelected ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SelectItemIndicatorFrame, { "aria-hidden": true, ...itemIndicatorProps, ref: forwardedRef }) : null;
  }
);
SelectItemIndicator.displayName = ITEM_INDICATOR_NAME;
const GROUP_NAME = "SelectGroup";
const [SelectGroupContextProvider, useSelectGroupContext] = (0, import_context.createSelectContext)(GROUP_NAME);
const SelectGroupFrame = (0, import_core.styled)(import_stacks.YStack, {
  name: GROUP_NAME,
  width: "100%"
});
const NativeSelectTextFrame = (0, import_core.styled)(import_text.SizableText, {
  tag: "select",
  backgroundColor: "$background",
  borderColor: "$borderColor",
  hoverStyle: {
    backgroundColor: "$backgroundHover"
  }
});
const NativeSelectFrame = (0, import_core.styled)(import_stacks.ThemeableStack, {
  name: "NativeSelect",
  bordered: true,
  userSelect: "none",
  outlineWidth: 0,
  paddingRight: 10,
  variants: {
    size: {
      "...size": (val, extras) => {
        const { tokens } = extras;
        const paddingHorizontal = (0, import_core.getVariableValue)(tokens.space[val]);
        return {
          borderRadius: tokens.radius[val] ?? val,
          minHeight: tokens.size[val],
          paddingRight: paddingHorizontal + 20,
          paddingLeft: paddingHorizontal,
          paddingVertical: (0, import_get_token.getSpace)(val, {
            shift: -3
          })
        };
      }
    }
  },
  defaultVariants: {
    size: "$2"
  }
});
const SelectGroup = React.forwardRef(
  (props, forwardedRef) => {
    const { __scopeSelect, ...groupProps } = props;
    const groupId = React.useId();
    const context = (0, import_context.useSelectContext)(GROUP_NAME, __scopeSelect);
    const itemParentContext = (0, import_context.useSelectItemParentContext)(GROUP_NAME, __scopeSelect);
    const size = itemParentContext.size ?? "$true";
    const nativeSelectRef = React.useRef(null);
    const content = function() {
      if (itemParentContext.shouldRenderWebNative) {
        return (
          // @ts-expect-error until we support typing based on tag
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(NativeSelectFrame, { asChild: true, size, value: context.value, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            NativeSelectTextFrame,
            {
              onChange: (event) => {
                itemParentContext.onChange(event.currentTarget.value);
              },
              size,
              ref: nativeSelectRef,
              style: {
                color: "var(--color)",
                // @ts-ignore
                appearance: "none"
              },
              children: props.children
            }
          ) })
        );
      }
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        SelectGroupFrame,
        {
          role: "group",
          "aria-labelledby": groupId,
          ...groupProps,
          ref: forwardedRef
        }
      );
    }();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SelectGroupContextProvider, { scope: __scopeSelect, id: groupId || "", children: content });
  }
);
SelectGroup.displayName = GROUP_NAME;
const LABEL_NAME = "SelectLabel";
const SelectLabel = React.forwardRef(
  (props, forwardedRef) => {
    const { __scopeSelect, ...labelProps } = props;
    const context = (0, import_context.useSelectItemParentContext)(LABEL_NAME, __scopeSelect);
    const groupContext = useSelectGroupContext(LABEL_NAME, __scopeSelect);
    if (context.shouldRenderWebNative) {
      return null;
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_list_item.ListItem,
      {
        tag: "div",
        componentName: LABEL_NAME,
        fontWeight: "800",
        id: groupContext.id,
        size: context.size,
        ...labelProps,
        ref: forwardedRef
      }
    );
  }
);
SelectLabel.displayName = LABEL_NAME;
const SelectSeparator = (0, import_core.styled)(import_separator.Separator, {
  name: "SelectSeparator"
});
const SelectSheetController = (props) => {
  const context = (0, import_context.useSelectContext)("SelectSheetController", props.__scopeSelect);
  const showSheet = (0, import_useSelectBreakpointActive.useShowSelectSheet)(context);
  const breakpointActive = (0, import_useSelectBreakpointActive.useSelectBreakpointActive)(context.sheetBreakpoint);
  const getShowSheet = (0, import_core.useGet)(showSheet);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_sheet.SheetController,
    {
      onOpenChange: (val) => {
        if (getShowSheet()) {
          props.onOpenChange(val);
        }
      },
      open: context.open,
      hidden: breakpointActive === false,
      children: props.children
    }
  );
};
const SelectSheetImpl = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: props.children });
};
const Select = (0, import_core.withStaticProperties)(
  (props) => {
    const {
      __scopeSelect,
      native,
      children,
      open: openProp,
      defaultOpen,
      onOpenChange,
      value: valueProp,
      defaultValue,
      onValueChange,
      size: sizeProp = "$true",
      dir
    } = props;
    const id = React.useId();
    const scopeKey = __scopeSelect ? Object.keys(__scopeSelect)[0] ?? id : id;
    const { when, AdaptProvider } = (0, import_adapt.useAdaptParent)({
      Contents: React.useCallback(
        () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_portal.PortalHost, { name: `${scopeKey}SheetContents` }),
        [scopeKey]
      )
    });
    const sheetBreakpoint = when;
    const isSheet = (0, import_useSelectBreakpointActive.useSelectBreakpointActive)(sheetBreakpoint);
    const SelectImpl = isSheet || !import_core.isWeb ? SelectSheetImpl : import_SelectImpl.SelectInlineImpl;
    const forceUpdate = React.useReducer(() => ({}), {})[1];
    const [selectedItem, setSelectedItem] = React.useState(null);
    const [open, setOpen] = (0, import_use_controllable_state.useControllableState)({
      prop: openProp,
      defaultProp: defaultOpen || false,
      onChange: onOpenChange
    });
    const [value, setValue] = (0, import_use_controllable_state.useControllableState)({
      prop: valueProp,
      defaultProp: defaultValue || "",
      onChange: onValueChange,
      transition: true
    });
    React.useEffect(() => {
      if (open) {
        emitValue(value);
      }
    }, [open]);
    const [activeIndex, setActiveIndex] = React.useState(0);
    const [emitValue, valueSubscribe] = useEmitter();
    const [emitActiveIndex, activeIndexSubscribe] = useEmitter();
    const selectedIndexRef = React.useRef(null);
    const activeIndexRef = React.useRef(null);
    const listContentRef = React.useRef([]);
    const [selectedIndex, setSelectedIndex] = React.useState(0);
    const [valueNode, setValueNode] = React.useState(null);
    (0, import_core.useIsomorphicLayoutEffect)(() => {
      selectedIndexRef.current = selectedIndex;
      activeIndexRef.current = activeIndex;
    });
    const shouldRenderWebNative = import_core.isWeb && (native === true || native === "web" || Array.isArray(native) && native.includes("web"));
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(AdaptProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_context.SelectItemParentProvider,
      {
        scope: __scopeSelect,
        initialValue: React.useMemo(() => value, []),
        size: sizeProp,
        activeIndexSubscribe,
        valueSubscribe,
        setOpen,
        onChange: React.useCallback((val) => {
          setValue(val);
          emitValue(val);
        }, []),
        setSelectedIndex,
        setValueAtIndex: React.useCallback((index, value2) => {
          listContentRef.current[index] = value2;
        }, []),
        shouldRenderWebNative,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_context.SelectProvider,
          {
            scope: __scopeSelect,
            dir,
            blockSelection: false,
            fallback: false,
            selectedItem,
            setSelectedItem,
            forceUpdate,
            valueNode,
            onValueNodeChange: setValueNode,
            scopeKey,
            sheetBreakpoint,
            activeIndex,
            selectedIndex,
            setActiveIndex: React.useCallback((index) => {
              setActiveIndex(index);
              if (typeof index === "number") {
                emitActiveIndex(index);
              }
            }, []),
            value,
            open,
            native,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SelectSheetController, { onOpenChange: setOpen, __scopeSelect, children: shouldRenderWebNative ? children : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              SelectImpl,
              {
                activeIndexRef,
                listContentRef,
                selectedIndexRef,
                ...props,
                open,
                value,
                children
              }
            ) })
          }
        )
      }
    ) });
  },
  {
    Adapt: import_adapt.Adapt,
    Content: import_SelectContent.SelectContent,
    Group: SelectGroup,
    Icon: SelectIcon,
    Item: import_SelectItem.SelectItem,
    ItemIndicator: SelectItemIndicator,
    ItemText: SelectItemText,
    Label: SelectLabel,
    ScrollDownButton: import_SelectScrollButton.SelectScrollDownButton,
    ScrollUpButton: import_SelectScrollButton.SelectScrollUpButton,
    Trigger: import_SelectTrigger.SelectTrigger,
    Value: SelectValue,
    Viewport: import_SelectViewport.SelectViewport,
    Sheet: import_sheet.Sheet.Controlled
  }
);
function useEmitter() {
  const listeners = React.useRef();
  if (!listeners.current) {
    listeners.current = /* @__PURE__ */ new Set();
  }
  const emit = (value) => {
    listeners.current.forEach((l) => l(value));
  };
  const subscribe = React.useCallback((listener) => {
    listeners.current.add(listener);
    return () => {
      listeners.current.delete(listener);
    };
  }, []);
  return [emit, subscribe];
}
Select.displayName = import_constants.SELECT_NAME;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Select,
  SelectGroupFrame,
  SelectIcon,
  SelectItemTextFrame,
  SelectSeparator
});
