// @tamagui/web/dist/cjs/constants/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  THEME_CLASSNAME_PREFIX: () => THEME_CLASSNAME_PREFIX,
  THEME_NAME_SEPARATOR: () => THEME_NAME_SEPARATOR,
  stackDefaultStyles: () => stackDefaultStyles
});
module.exports = __toCommonJS(constants_exports);
const THEME_NAME_SEPARATOR = "_";
const THEME_CLASSNAME_PREFIX = "t_";
const stackDefaultStyles = {
  alignItems: "stretch"
};
if (process.env.TAMAGUI_TARGET === "web") {
  Object.assign(stackDefaultStyles, {
    display: "flex",
    flexDirection: "column",
    flexBasis: "auto",
    boxSizing: "border-box",
    position: "relative",
    minHeight: 0,
    minWidth: 0,
    flexShrink: 0
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  THEME_CLASSNAME_PREFIX,
  THEME_NAME_SEPARATOR,
  stackDefaultStyles
});
