// @tamagui/toast/dist/cjs/Toast.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Toast_exports = {};
__export(Toast_exports, {
  Toast: () => Toast,
  ToastProvider: () => import_ToastProvider.ToastProvider,
  ToastViewport: () => import_ToastViewport.ToastViewport,
  createToastScope: () => import_ToastProvider.createToastScope,
  useToast: () => import_ToastImperative.useToast,
  useToastController: () => import_ToastImperative.useToastController,
  useToastState: () => import_ToastImperative.useToastState
});
module.exports = __toCommonJS(Toast_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@tamagui/core");
var import_stacks = require("@tamagui/stacks");
var import_text = require("@tamagui/text");
var import_use_controllable_state = require("@tamagui/use-controllable-state");
var React = __toESM(require("react"));
var import_constants = require("./constants");
var import_ToastAnnounce = require("./ToastAnnounce");
var import_ToastImperative = require("./ToastImperative");
var import_ToastImpl = require("./ToastImpl");
var import_ToastProvider = require("./ToastProvider");
var import_ToastViewport = require("./ToastViewport");
const TITLE_NAME = "ToastTitle";
const ToastTitle = (0, import_core.styled)(import_text.SizableText, {
  name: TITLE_NAME,
  variants: {
    unstyled: {
      false: {
        color: "$color",
        size: "$4"
      }
    }
  },
  defaultVariants: {
    unstyled: false
  }
});
ToastTitle.displayName = TITLE_NAME;
const DESCRIPTION_NAME = "ToastDescription";
const ToastDescription = (0, import_core.styled)(import_text.SizableText, {
  name: DESCRIPTION_NAME,
  variants: {
    unstyled: {
      false: {
        color: "$color11",
        size: "$1"
      }
    }
  },
  defaultVariants: {
    unstyled: false
  }
});
ToastDescription.displayName = DESCRIPTION_NAME;
const ACTION_NAME = "ToastAction";
const ToastAction = React.forwardRef(
  (props, forwardedRef) => {
    const { altText, ...actionProps } = props;
    if (!altText)
      return null;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ToastAnnounce.ToastAnnounceExclude, { altText, asChild: true, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ToastClose, { ...actionProps, ref: forwardedRef }) });
  }
);
ToastAction.propTypes = {
  altText(props) {
    if (!props.altText) {
      return new Error(`Missing prop \`altText\` expected on \`${ACTION_NAME}\``);
    }
    return null;
  }
};
ToastAction.displayName = ACTION_NAME;
const CLOSE_NAME = "ToastClose";
const ToastCloseFrame = (0, import_core.styled)(import_stacks.ThemeableStack, {
  name: CLOSE_NAME,
  tag: "button"
});
const ToastClose = React.forwardRef(
  (props, forwardedRef) => {
    const { __scopeToast, ...closeProps } = props;
    const interactiveContext = (0, import_ToastImpl.useToastInteractiveContext)(CLOSE_NAME, __scopeToast);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ToastAnnounce.ToastAnnounceExclude, { asChild: true, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      ToastCloseFrame,
      {
        accessibilityLabel: "Dialog Close",
        ...closeProps,
        ref: forwardedRef,
        onPress: (0, import_core.composeEventHandlers)(props.onPress, interactiveContext.onClose)
      }
    ) });
  }
);
ToastClose.displayName = CLOSE_NAME;
const ToastComponent = React.forwardRef(
  import_ToastImpl.ToastImplFrame.extractable((props, forwardedRef) => {
    const { forceMount, open: openProp, defaultOpen, onOpenChange, ...toastProps } = props;
    const [open, setOpen] = (0, import_use_controllable_state.useControllableState)({
      prop: openProp,
      defaultProp: defaultOpen ?? true,
      onChange: onOpenChange,
      strategy: "most-recent-wins"
    });
    const id = React.useId();
    const onPause = (0, import_core.useEvent)(props.onPause);
    const onResume = (0, import_core.useEvent)(props.onResume);
    const shouldShow = forceMount || open;
    if (!shouldShow)
      return null;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ToastImpl.ToastImpl,
      {
        id,
        open,
        ...toastProps,
        ref: forwardedRef,
        onClose: () => setOpen(false),
        onPause,
        onResume,
        onSwipeEnd: (0, import_core.composeEventHandlers)(props.onSwipeEnd, (event) => {
          setOpen(false);
        })
      }
    );
  })
);
ToastComponent.displayName = import_constants.TOAST_NAME;
const Toast = (0, import_core.withStaticProperties)(
  (0, import_core.themeable)(ToastComponent, import_ToastImpl.ToastImplFrame.staticConfig),
  {
    Title: ToastTitle,
    Description: ToastDescription,
    Action: ToastAction,
    Close: ToastClose
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Toast,
  ToastProvider,
  ToastViewport,
  createToastScope,
  useToast,
  useToastController,
  useToastState
});
