// @tamagui/react-native-svg/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  Circle: () => Circle,
  ClipPath: () => ClipPath,
  Defs: () => Defs,
  Ellipse: () => Ellipse,
  ForeignObject: () => ForeignObject,
  G: () => G,
  Image: () => Image,
  Line: () => Line,
  LinearGradient: () => LinearGradient,
  Marker: () => Marker,
  Mask: () => Mask,
  Path: () => Path,
  Pattern: () => Pattern,
  Polygon: () => Polygon,
  Polyline: () => Polyline,
  RadialGradient: () => RadialGradient,
  Rect: () => Rect,
  Shape: () => Shape,
  Stop: () => Stop,
  Svg: () => Svg,
  Symbol: () => Symbol2,
  TSpan: () => TSpan,
  Text: () => Text,
  TextPath: () => TextPath,
  Use: () => Use,
  default: () => src_default
});
module.exports = __toCommonJS(src_exports);
var import_react = require("react");
const Circle = (p) => (0, import_react.createElement)("circle", p);
const ClipPath = (p) => (0, import_react.createElement)("clipPath", p);
const Defs = (p) => (0, import_react.createElement)("defs", p);
const Ellipse = (p) => (0, import_react.createElement)("ellipse", p);
const ForeignObject = (p) => (0, import_react.createElement)("foreignObject", p);
const G = (p) => (0, import_react.createElement)("g", p);
const Image = (p) => (0, import_react.createElement)("image", p);
const Line = (p) => (0, import_react.createElement)("line", p);
const LinearGradient = (p) => (0, import_react.createElement)("linearGradient", p);
const Marker = (p) => (0, import_react.createElement)("marker", p);
const Mask = (p) => (0, import_react.createElement)("mask", p);
const Path = (p) => (0, import_react.createElement)("path", p);
const Pattern = (p) => (0, import_react.createElement)("pattern", p);
const Polygon = (p) => (0, import_react.createElement)("polygon", p);
const Polyline = (p) => (0, import_react.createElement)("polyline", p);
const RadialGradient = (p) => (0, import_react.createElement)("radialGradient", p);
const Rect = (p) => (0, import_react.createElement)("rect", p);
const Shape = (p) => (0, import_react.createElement)("shape", p);
const Stop = (p) => (0, import_react.createElement)("stop", p);
const Svg = (p) => (0, import_react.createElement)("svg", p);
const Symbol2 = (p) => (0, import_react.createElement)("symbol", p);
const Text = (p) => (0, import_react.createElement)("text", p);
const TextPath = (p) => (0, import_react.createElement)("clipPath", p);
const TSpan = (p) => (0, import_react.createElement)("tSpan", p);
const Use = (p) => (0, import_react.createElement)("use", p);
var src_default = {
  Circle,
  ClipPath,
  Defs,
  Ellipse,
  ForeignObject,
  G,
  Image,
  Line,
  LinearGradient,
  Marker,
  Mask,
  Path,
  Pattern,
  Polygon,
  Polyline,
  RadialGradient,
  Rect,
  Shape,
  Stop,
  Svg,
  Symbol: Symbol2,
  Text,
  TextPath,
  TSpan,
  Use
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Circle,
  ClipPath,
  Defs,
  Ellipse,
  ForeignObject,
  G,
  Image,
  Line,
  LinearGradient,
  Marker,
  Mask,
  Path,
  Pattern,
  Polygon,
  Polyline,
  RadialGradient,
  Rect,
  Shape,
  Stop,
  Svg,
  Symbol,
  TSpan,
  Text,
  TextPath,
  Use
});
