// @tamagui/static/dist/cjs/extractor/getStaticBindingsForScope.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var getStaticBindingsForScope_exports = {};
__export(getStaticBindingsForScope_exports, {
  cleanupBeforeExit: () => cleanupBeforeExit,
  getStaticBindingsForScope: () => getStaticBindingsForScope
});
module.exports = __toCommonJS(getStaticBindingsForScope_exports);
var import_child_process = require("child_process");
var import_path = require("path");
var t = __toESM(require("@babel/types"));
var import_evaluateAstNode = require("./evaluateAstNode");
var import_getSourceModule = require("./getSourceModule");
const isLocalImport = (path) => path.startsWith(".") || path.startsWith("/");
function resolveImportPath(sourcePath, path) {
  const sourceDir = (0, import_path.dirname)(sourcePath);
  if (isLocalImport(path)) {
    if ((0, import_path.extname)(path) === "") {
      path += ".js";
    }
    return (0, import_path.resolve)(sourceDir, path);
  }
  return path;
}
const cache = /* @__PURE__ */ new Map();
const pending = /* @__PURE__ */ new Map();
const loadCmd = `${(0, import_path.join)(__dirname, "loadFile.js")}`;
let exited = false;
let child = null;
function forkChild() {
  child = (0, import_child_process.fork)(loadCmd, [], {
    execArgv: ["-r", "esbuild-register"],
    detached: false,
    stdio: "ignore"
  });
}
function cleanupBeforeExit() {
  if (exited)
    return;
  if (!child)
    return;
  child.removeAllListeners();
  child.unref();
  child.disconnect();
  child.kill();
  exited = true;
}
process.once("SIGTERM", cleanupBeforeExit);
process.once("SIGINT", cleanupBeforeExit);
process.once("beforeExit", cleanupBeforeExit);
function importModule(path) {
  if (!child) {
    forkChild();
  }
  if (pending.has(path)) {
    return pending.get(path);
  }
  const promise = new Promise((res, rej) => {
    if (!child)
      return;
    if (cache.size > 2e3) {
      cache.clear();
    }
    if (cache.has(path)) {
      return cache.get(path);
    }
    const listener = (msg) => {
      if (!child)
        return;
      if (!msg)
        return;
      if (typeof msg !== "string")
        return;
      if (msg[0] === "-") {
        rej(new Error(msg.slice(1)));
        return;
      }
      child.removeListener("message", listener);
      const val = JSON.parse(msg);
      cache.set(path, val);
      res(val);
    };
    child.once("message", listener);
    child.send(`${path.replace(".js", "")}`);
  });
  pending.set(path, promise);
  return promise;
}
async function getStaticBindingsForScope(scope, whitelist = [], sourcePath, bindingCache, shouldPrintDebug) {
  var _a;
  const bindings = scope.getAllBindings();
  const ret = {};
  if (shouldPrintDebug) {
  }
  const program = scope.getProgramParent().block;
  for (const node of program.body) {
    if (t.isImportDeclaration(node)) {
      const importPath = node.source.value;
      if (!node.specifiers.length)
        continue;
      if (!isLocalImport(importPath)) {
        continue;
      }
      const moduleName = resolveImportPath(sourcePath, importPath);
      const isOnWhitelist = whitelist.some((test) => moduleName.endsWith(test));
      if (!isOnWhitelist)
        continue;
      try {
        const src = await importModule(moduleName);
        if (!src)
          continue;
        for (const specifier of node.specifiers) {
          if (t.isImportSpecifier(specifier) && t.isIdentifier(specifier.imported)) {
            if (typeof src[specifier.imported.name] !== "undefined") {
              const val = src[specifier.local.name];
              ret[specifier.local.name] = val;
            }
          }
        }
      } catch (err) {
        if (shouldPrintDebug) {
          console.warn(
            `    | Skipping partial evaluation of constant file: ${moduleName} (DEBUG=tamagui for more)`
          );
        } else if ((_a = process.env.DEBUG) == null ? void 0 : _a.startsWith("tamagui")) {
          console.log(`Error in partial evaluation`, err.message, err.stack);
        }
      }
    }
  }
  if (!bindingCache) {
    throw new Error("BindingCache is a required param");
  }
  for (const k in bindings) {
    const binding = bindings[k];
    const sourceModule = (0, import_getSourceModule.getSourceModule)(k, binding);
    if (sourceModule) {
      if (!sourceModule.sourceModule) {
        continue;
      }
      const moduleName = resolveImportPath(sourcePath, sourceModule.sourceModule);
      const isOnWhitelist = whitelist.some((test) => moduleName.endsWith(test));
      if (isOnWhitelist) {
        const src = importModule(moduleName);
        if (!src) {
          console.log(
            `    | \u26A0\uFE0F Missing file ${moduleName} via ${sourcePath} import ${sourceModule.sourceModule}?`
          );
          return {};
        }
        if (sourceModule.destructured) {
          if (sourceModule.imported) {
            ret[k] = src[sourceModule.imported];
          }
        } else {
          ret[k] = src;
        }
      }
      continue;
    }
    const { parent } = binding.path;
    if (!t.isVariableDeclaration(parent) || parent.kind !== "const") {
      continue;
    }
    const dec = parent.declarations.find((d) => t.isIdentifier(d.id) && d.id.name === k);
    if (!dec || !dec.init) {
      continue;
    }
    if (typeof dec.id.start !== "number" || typeof dec.id.end !== "number") {
      console.error("dec.id.start/end is not a number");
      continue;
    }
    if (!t.isIdentifier(dec.id)) {
      console.error("dec is not an identifier");
      continue;
    }
    const cacheKey = `${dec.id.name}_${dec.id.start}-${dec.id.end}`;
    if (Object.hasOwnProperty.call(bindingCache, cacheKey)) {
      ret[k] = bindingCache[cacheKey];
      continue;
    }
    if (Object.hasOwnProperty.call(bindingCache, cacheKey)) {
      ret[k] = bindingCache[cacheKey];
      continue;
    }
    try {
      ret[k] = (0, import_evaluateAstNode.evaluateAstNode)(dec.init, void 0, shouldPrintDebug);
      bindingCache[cacheKey] = ret[k];
      continue;
    } catch {
    }
  }
  return ret;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  cleanupBeforeExit,
  getStaticBindingsForScope
});
