// @tamagui/static/dist/cjs/extractor/validHTMLAttributes.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var validHTMLAttributes_exports = {};
__export(validHTMLAttributes_exports, {
  validHTMLAttributes: () => validHTMLAttributes
});
module.exports = __toCommonJS(validHTMLAttributes_exports);
const validHTMLAttributes = {
  autocomplete: true,
  border: true,
  contenteditable: true,
  crossorigin: true,
  dir: true,
  draggable: true,
  enctype: true,
  formenctype: true,
  formmethod: true,
  formtarget: true,
  inputmode: true,
  kind: true,
  link: true,
  method: true,
  preload: true,
  referrerpolicy: true,
  rel: true,
  rev: true,
  role: true,
  sandbox: true,
  shape: true,
  spellcheck: true,
  target: true,
  translate: true,
  type: true,
  wrap: true,
  "aria-autocomplete": true,
  "aria-busy": true,
  "aria-checked": true,
  "aria-current": true,
  "aria-disabled": true,
  "aria-expanded": true,
  "aria-haspopup": true,
  "aria-hidden": true,
  "aria-invalid": true,
  "aria-polite": true,
  "aria-modal": true,
  "aria-multiline": true,
  "aria-multiselectable": true,
  "aria-orientation": true,
  "aria-pressed": true,
  "aria-readonly": true,
  "aria-relevant": true,
  "aria-required": true,
  "aria-selected": true,
  "aria-sort": true
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  validHTMLAttributes
});
