// @tamagui/sheet/dist/cjs/createSheet.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var createSheet_exports = {};
__export(createSheet_exports, {
  createSheet: () => createSheet
});
module.exports = __toCommonJS(createSheet_exports);
var import_jsx_runtime = (
  // @ts-ignore
  require("react/jsx-runtime")
);
var import_compose_refs = require("@tamagui/compose-refs");
var import_core = require("@tamagui/core");
var import_remove_scroll = require("@tamagui/remove-scroll");
var import_react = require("react");
var import_react_native = require("react-native");
var import_constants = require("./constants");
var import_nativeSheet = require("./nativeSheet");
var import_SheetContext = require("./SheetContext");
var import_SheetImplementationCustom = require("./SheetImplementationCustom");
var import_SheetScrollView = require("./SheetScrollView");
var import_useSheetController = require("./useSheetController");
var import_useSheetOffscreenSize = require("./useSheetOffscreenSize");
function createSheet({ Handle, Frame, Overlay }) {
  const SheetHandle = Handle.extractable(
    ({ __scopeSheet, ...props }) => {
      const context = (0, import_SheetContext.useSheetContext)(import_constants.SHEET_HANDLE_NAME, __scopeSheet);
      if (context.onlyShowFrame) {
        return null;
      }
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        Handle,
        {
          onPress: () => {
            const max = context.snapPoints.length + (context.dismissOnSnapToBottom ? -1 : 0);
            const nextPos = (context.position + 1) % max;
            context.setPosition(nextPos);
          },
          open: context.open,
          ...props
        }
      );
    }
  );
  const SheetOverlay = Overlay.extractable(
    (0, import_react.memo)((propsIn) => {
      const { __scopeSheet, ...props } = propsIn;
      const context = (0, import_SheetContext.useSheetContext)(import_constants.SHEET_OVERLAY_NAME, __scopeSheet);
      const element = (0, import_react.useMemo)(
        () => (
          // @ts-ignore
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            Overlay,
            {
              ...props,
              onPress: (0, import_core.composeEventHandlers)(
                props.onPress,
                context.dismissOnOverlayPress ? () => {
                  context.setOpen(false);
                } : void 0
              )
            }
          )
        ),
        [props.onPress, context.dismissOnOverlayPress]
      );
      (0, import_core.useIsomorphicLayoutEffect)(() => {
        var _a;
        (_a = context.onOverlayComponent) == null ? void 0 : _a.call(context, element);
      }, [element]);
      if (context.onlyShowFrame) {
        return null;
      }
      return null;
    })
  );
  const SheetFrame = Frame.extractable(
    (0, import_react.forwardRef)(
      ({
        __scopeSheet,
        children,
        ...props
      }, forwardedRef) => {
        const context = (0, import_SheetContext.useSheetContext)(import_constants.SHEET_NAME, __scopeSheet);
        const { removeScrollEnabled, frameSize, contentRef } = context;
        const composedContentRef = (0, import_compose_refs.useComposedRefs)(forwardedRef, contentRef);
        const offscreenSize = (0, import_useSheetOffscreenSize.useSheetOffscreenSize)(context);
        const handleLayoutChange = (0, import_react.useCallback)((e) => {
          var _a;
          const next = (_a = e.nativeEvent) == null ? void 0 : _a.layout.height;
          if (!next)
            return;
          context.setFrameSize(next);
        }, []);
        const sheetContents = (0, import_react.useMemo)(() => {
          return (
            // @ts-ignore
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              Frame,
              {
                ref: composedContentRef,
                height: context.hasFit ? void 0 : frameSize,
                onLayout: context.hasFit ? handleLayoutChange : void 0,
                ...props,
                children: [
                  children,
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_core.Stack, { "data-sheet-offscreen-pad": true, height: offscreenSize, width: "100%" })
                ]
              }
            )
          );
        }, [props, frameSize, offscreenSize]);
        return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_remove_scroll.RemoveScroll,
            {
              forwardProps: true,
              enabled: removeScrollEnabled,
              allowPinchZoom: true,
              shards: [contentRef],
              removeScrollBar: false,
              children: sheetContents
            }
          ),
          !props.disableHideBottomOverflow && // @ts-ignore
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            Frame,
            {
              ...props,
              componentName: "SheetCover",
              children: null,
              position: "absolute",
              bottom: "-50%",
              zIndex: -1,
              height: context.frameSize,
              left: 0,
              right: 0,
              borderWidth: 0,
              borderRadius: 0,
              shadowOpacity: 0
            }
          )
        ] });
      }
    )
  );
  const Sheet = (0, import_react.forwardRef)(function Sheet2(props, ref) {
    const hydrated = (0, import_core.useDidFinishSSR)();
    const { isShowingNonSheet } = (0, import_useSheetController.useSheetController)();
    let SheetImplementation = import_SheetImplementationCustom.SheetImplementationCustom;
    if (props.native && import_react_native.Platform.OS === "ios") {
      if (process.env.TAMAGUI_TARGET === "native") {
        const impl = (0, import_nativeSheet.getNativeSheet)("ios");
        if (impl) {
          SheetImplementation = impl;
        }
      }
    }
    if (isShowingNonSheet || !hydrated) {
      return null;
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SheetImplementation, { ref, ...props });
  });
  const components = {
    Frame: SheetFrame,
    Overlay: SheetOverlay,
    Handle: SheetHandle,
    ScrollView: import_SheetScrollView.SheetScrollView
  };
  const Controlled = (0, import_core.withStaticProperties)(Sheet, components);
  return (0, import_core.withStaticProperties)(Sheet, {
    ...components,
    Controlled
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createSheet
});
