// @tamagui/sheet/dist/cjs/Sheet.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Sheet_exports = {};
__export(Sheet_exports, {
  Frame: () => Frame,
  Handle: () => Handle,
  Overlay: () => Overlay,
  Sheet: () => Sheet,
  SheetHandleFrame: () => SheetHandleFrame,
  SheetOverlayFrame: () => SheetOverlayFrame,
  createSheetScope: () => import_SheetContext.createSheetScope
});
module.exports = __toCommonJS(Sheet_exports);
var import_core = require("@tamagui/core");
var import_stacks = require("@tamagui/stacks");
var import_constants = require("./constants");
var import_createSheet = require("./createSheet");
var import_SheetContext = require("./SheetContext");
__reExport(Sheet_exports, require("./types"), module.exports);
const Handle = (0, import_core.styled)(import_stacks.XStack, {
  name: import_constants.SHEET_HANDLE_NAME,
  variants: {
    open: {
      true: {
        pointerEvents: "auto"
      },
      false: {
        opacity: 0,
        pointerEvents: "none"
      }
    },
    unstyled: {
      false: {
        height: 10,
        borderRadius: 100,
        backgroundColor: "$background",
        zIndex: 10,
        marginHorizontal: "35%",
        marginBottom: "$2",
        opacity: 0.5,
        hoverStyle: {
          opacity: 0.7
        }
      }
    }
  },
  defaultVariants: {
    unstyled: false
  }
});
const Overlay = (0, import_core.styled)(import_stacks.ThemeableStack, {
  name: import_constants.SHEET_OVERLAY_NAME,
  variants: {
    open: {
      true: {
        opacity: 1,
        pointerEvents: "auto"
      },
      false: {
        opacity: 0,
        pointerEvents: "none"
      }
    },
    unstyled: {
      false: {
        fullscreen: true,
        position: "absolute",
        backgrounded: true,
        zIndex: 1e5 - 1,
        pointerEvents: "auto"
      }
    }
  },
  defaultVariants: {
    unstyled: false
  }
});
const Frame = (0, import_core.styled)(import_stacks.YStack, {
  name: import_constants.SHEET_NAME,
  variants: {
    unstyled: {
      false: {
        flex: 1,
        backgroundColor: "$background",
        borderTopLeftRadius: "$true",
        borderTopRightRadius: "$true",
        width: "100%",
        maxHeight: "100%",
        overflow: "hidden"
      }
    }
  },
  defaultVariants: {
    unstyled: false
  }
});
const Sheet = (0, import_createSheet.createSheet)({
  Frame,
  Handle,
  Overlay
});
const SheetOverlayFrame = Overlay;
const SheetHandleFrame = Handle;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Frame,
  Handle,
  Overlay,
  Sheet,
  SheetHandleFrame,
  SheetOverlayFrame,
  createSheetScope,
  ...require("./types")
});
