// @tamagui/portal/dist/cjs/Portal.native.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Portal_native_exports = {};
__export(Portal_native_exports, {
  Portal: () => Portal
});
module.exports = __toCommonJS(Portal_native_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_stacks = require("@tamagui/stacks");
var React = __toESM(require("react"));
var import_react_native = require("react-native");
var import_ReactNative = require("react-native/Libraries/Renderer/shims/ReactNative");
var import_GorhomPortal = require("./GorhomPortal");
const Portal = (props) => {
  const rootTag = React.useContext(import_react_native.RootTagContext);
  const contents = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_stacks.YStack,
    {
      pointerEvents: "box-none",
      fullscreen: true,
      position: "absolute",
      maxWidth: "100%",
      zIndex: 1e5,
      ...props
    }
  );
  if (import_react_native.Platform.OS === "android" || !rootTag) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GorhomPortal.PortalItem, { hostName: "root", children: contents });
  }
  return (0, import_ReactNative.createPortal)(contents, rootTag);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Portal
});
